contracts/core/OriginCore.sol

```diff
@@ -144,43 +112,110 @@ contract OriginCore is OriginCoreInterface, OriginCoreConfig {
      * @param _coreIdentifier A unique identifier that identifies what chain
      *                        this vote is about.
-     * @param _auxiliaryBlockHash The hash of the last finalised checkpoint
-     *                            that is part of this meta-block.
-     * @param _gas The total consumed gas on auxiliary within this meta-block.
+     * @param _kernelHash The hash of the current kernel.
+     * @param _auxiliaryDynasty The dynasty number where the meta-block closes
+     *                          on the auxiliary chain.
+     * @param _auxiliaryBlockHash The block hash where the meta-block closes
+     *                          on the auxiliary chain.
+     * @param _accumulatedGas The total consumed gas on auxiliary within
+     *                        this meta-block.
+     * @param _originDynasty Dynasty of origin block within latest meta-block
+     *                          reported at auxiliary chain.
+     * @param _originBlockHash Block hash of origin block within latest
+     *                          meta-block reported at auxiliary chain.
      * @param _transactionRoot The transaction root of the meta-block. A trie
      *                         created by the auxiliary block store from the
      *                         transaction roots of all blocks.
-     * @param _auxiliaryDynasty The dynasty number where the meta-block closes
-     *                          on the auxiliary chain.
-     *
      * @return `true` if the proposal succeeds.
      */
     function proposeBlock(
         uint256 _height,
-        bytes32 _parent,
-        address[] _updatedValidators,
-        uint256[] _updatedWeights,
-        bytes20 _coreIdentifier,
+        bytes32 _coreIdentifier,
+        bytes32 _kernelHash,
+        uint256 _auxiliaryDynasty,
         bytes32 _auxiliaryBlockHash,
-        uint256 _gas,
-        bytes32 _transactionRoot,
-        uint256 _auxiliaryDynasty
+        uint256 _accumulatedGas,
+        uint256 _originDynasty,
+        bytes32 _originBlockHash,
+        bytes32 _transactionRoot
     )
         external
-        returns (bool success_)
+        returns (bool)
     {
-        revert("Method not implemented.");
+
+        require(
+            _kernelHash != bytes32(0),
+            'Kernel hash should not be `0`.'
+        );
+
+        require(
+            _originDynasty > 0,
+            'Origin dynasty should not be `0`.'
+        );
+
+        require(
+            _originBlockHash != bytes32(0),
+            'Origin block should not be `0`.'
+        );
+
+        require(
+            _transactionRoot != bytes32(0),
+            'Transaction Root hash should not be `0`.'
+        );
+
+        require(
+            _coreIdentifier == auxiliaryCoreIdentifier,
+            'CoreIdentifier should be same as auxiliary core identifier.'
+        );
+
+        /* header of last meta block */
+        MetaBlock.Header latestMetaBlockHeader = reportedHeaders[head];
```

Comment:
You should explicitly name the data location. It makes the code easier to read. It will also be mandatory in [solidity 0.5](https://solidity.readthedocs.io/en/develop/types.html#data-location):
>Prior to version 0.5.0 the data location could be omitted, and would default to different locations depending on the kind of variable, function type, etc., but all complex types must now give an explicit data location.

```suggestion
        MetaBlock.Header storage latestMetaBlockHeader = reportedHeaders[head];
```

Reply:
Sure, I  will do this. We should add such points in the style guide?